### PR TITLE
GuardDog: Deadlock detection for stats/termination

### PR DESCRIPTION
**Pull request by dnoe**
This commit adds a deadlock detector called GuardDog. It replaces the
existing watchdog timer and performs the same statistic collecting
functionality while also providing the ability to detect and act on true
deadlocks (one or more event processing threads becoming permanently
nonresponsive). By performing all deadlock checking in an independent
thread we can PANIC() and kill the process for a semi-graceful restart
rather than hanging.  This should produce a core file which can be used
to find out details on the stuck threads.

New configuration variables are introduced to the top level
configuration to control the new kill behavior and also make the
existing thresholds for statistics incrementing configurable. The
configuration details are described in the configuration overview
document.

The new bits of implementation were placed into Event namespace because
they are most relevant to threads that are following the Event and
Dispatcher paradigm.  If this is not ideal it can be moved.

### CODE REVIEW

**Comment by mattklein123** [comment]
@htuch can you take first pass over this?

File: include/envoy/server/configuration.h [hunk 1]
```diff
@@ -132,6 +132,27 @@ class Main {
    *         counters.
    */
   virtual std::chrono::milliseconds statsFlushInterval() PURE;
+
+  /**
+   * @return the time interval after which we count a nonresponsive thread event
+   * as a "miss" statistic.
```
Comment (htuch) [inline include/envoy/server/configuration.h]: Nit: Add the return type at the beginning of the @return doc. This is more for consistency with other Envoy code.
Nit: Prefer space indenting on line wrap to underneath the first word of the previous line excluding @return.

File: source/common/event/guarddog_impl.cc [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+namespace Event {
+
+GuardDogImpl::GuardDogImpl(Stats::Store& stats_store, const Server::Configuration::Main& config,
+                           SystemTimeSource& tsource)
+    : time_source_(tsource), miss_timeout_(config.wdMissTimeout()),
+      megamiss_timeout_(config.wdMegaMissTimeout()), kill_timeout_(config.wdKillTimeout()),
+      multi_kill_timeout_(config.wdMultiKillTimeout()),
+      loop_interval_(
+          std::min({miss_timeout_, megamiss_timeout_,
+                    kill_enabled() ? kill_timeout_ : std::min(miss_timeout_, megamiss_timeout_),
```
Comment (htuch) [inline source/common/event/guarddog_impl.cc]: The repetition of `std::min(miss_timeout_, ..)` could be avoided if this was done inside the constructor rather than initializor list. You can simplify it a bit, since it's min and you just need something here >= miss_timeout_/megamiss_timeout_.

File: source/common/event/guarddog_impl.cc [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      watchdog_megamiss_counter_(stats_store.counter("server.watchdog_mega_miss")),
+      run_thread_(true) {
+  start();
+}
+
+GuardDogImpl::~GuardDogImpl() { exit(); }
+
+void GuardDogImpl::threadRoutine() {
+  do {
+    auto now = time_source_.currentSystemTime();
```
Comment (htuch) [inline source/common/event/guarddog_impl.cc]: Nit: const

File: source/common/event/guarddog_impl.cc [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    bool seen_one_multi_timeout(false);
+    std::lock_guard<std::mutex> guard(wd_lock_);
+    for (auto watched_dog : watched_dogs_) {
+      auto delta = now - watched_dog->lastTouchTime();
```
Comment (htuch) [inline source/common/event/guarddog_impl.cc]: Nit: const etc. Best to be as const as possible everywhere.

File: source/common/event/guarddog_impl.cc [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+          PANIC("GuardDog: multiple threads stuck for more than watchdog_multikill_timeout");
+        } else {
+          seen_one_multi_timeout = true;
+        }
+      }
+    }
+  } while (waitOrDetectStop());
+}
+
+WatchDogSharedPtr GuardDogImpl::getWatchDog(int32_t thread_id) {
```
Comment (htuch) [inline source/common/event/guarddog_impl.cc]: I think I prefer `createWatchDog`, since a new long lived object is being created here.

File: source/common/event/guarddog_impl.cc [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  // Timer started by WatchDog will try to fire at 3/4 of the interval of the
+  // minimum timeout specified.
+  auto wd_interval = loop_interval_ * 3 / 4;
```
Comment (htuch) [inline source/common/event/guarddog_impl.cc]: Where does 3/4 come from? Why not 1/2? I don't think it matters that much as long as it's some reasonable fraction, but a reader might..

File: source/common/event/guarddog_impl.cc [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  WatchDogSharedPtr new_watchdog =
+      std::make_shared<WatchDogImpl>(thread_id, time_source_, wd_interval);
+  {
+    std::lock_guard<std::mutex> guard(wd_lock_);
```
Comment (htuch) [inline source/common/event/guarddog_impl.cc]: Do you want to protect the entire method with `wd_lock_`, since it might be invoked by another thread? There's definitely shared variable access, e.g. `loop_interval_`, even if read-only. TSAN might get angry, but we might get a pass since it's only written at init. Dunno.

File: source/common/event/guarddog_impl.h [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+#include "envoy/server/configuration.h"
+#include "envoy/stats/stats.h"
+
+#include "common/common/logger.h"
+#include "common/common/thread.h"
+#include "common/event/libevent.h"
+
+namespace Event {
+
+/** This feature performs deadlock detection stats collection&enforcement.
```
Comment (htuch) [inline source/common/event/guarddog_impl.h]: Nit: new line, spaces.

File: source/common/event/guarddog_impl.h [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+  ~GuardDogImpl();
+
+  WatchDogSharedPtr getWatchDog(int32_t thread_id) override;
+  void stopWatching(WatchDogSharedPtr wd) override;
+
+  /**
+   * Exposed for testing purposes only (but harmless to call):
+   */
+  int loopInterval() const { return loop_interval_.count(); }
+  void force_check() {
```
Comment (htuch) [inline source/common/event/guarddog_impl.h]: Nit: `forceCheck`. I think you could also call these somethings like `loopIntervalCountforTest()`, `forceCheckForTest()` if they really are only ever to be used in testing.

File: source/common/event/guarddog_impl.h [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+class GuardDogImpl : public GuardDog {
+public:
+  /**
+   * @param stats_store Statistics store to write watchdog_miss and
+   * watchdog_mega_miss events into.
+   * @param config Configuration object.
+   *
+   * Configuration:
+   *   1) Miss timeout: Record a "miss" counter increment after this long has
+   *   passed.  Old default: 100ms.
```
Comment (htuch) [inline source/common/event/guarddog_impl.h]: Old default?

File: test/integration/fake_upstream.cc [hunk 11]
```diff
@@ -201,7 +201,7 @@ FakeUpstream::FakeUpstream(Ssl::ServerContext* ssl_ctx, uint32_t port,
 FakeUpstream::FakeUpstream(Ssl::ServerContext* ssl_ctx, Network::ListenSocketPtr&& listen_socket,
                            FakeHttpConnection::Type type)
     : ssl_ctx_(ssl_ctx), socket_(std::move(listen_socket)),
-      handler_(stats_store_, log(), Api::ApiPtr{new Api::Impl(std::chrono::milliseconds(10000))}),
+      handler_(log(), Api::ApiPtr{new Api::Impl(std::chrono::milliseconds(10000))}),
```
Comment (htuch) [inline test/integration/fake_upstream.cc]: Is this a related change?

File: source/server/worker.h [hunk 12]
```diff
@@ -15,13 +16,13 @@ typedef std::map<Server::Configuration::Listener*, Network::TcpListenSocketPtr>
  */
 class Worker : Logger::Loggable<Logger::Id::main> {
 public:
-  Worker(Stats::Store& stats_store, ThreadLocal::Instance& tls,
-         std::chrono::milliseconds file_flush_interval_msec);
+  Worker(ThreadLocal::Instance& tls, std::chrono::milliseconds file_flush_interval_msec);
```
Comment (htuch) [inline source/server/worker.h]: Related change?

File: test/common/event/guarddog_impl_test.cc [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+
+  testing::NiceMock<Server::Configuration::MockMain> config_miss;
+  testing::NiceMock<Server::Configuration::MockMain> config_mega;
+  testing::NiceMock<Stats::MockStore> stats_store;
+  MockSystemTimeSource time_source_;
+  std::chrono::system_clock::time_point time_point_;
+};
+
+TEST_F(GuardDogDeathTest, KillDeathTest) { EXPECT_DEATH(startNeglectedDogAndWait(), ""); }
+TEST_F(GuardDogDeathTest, MultiKillDeathTest) { EXPECT_DEATH(startNeglectedDogsAndWait(), ""); }
```
Comment (htuch) [inline test/common/event/guarddog_impl_test.cc]: See comment below about inlining stuff into the `TEST_F`. Also, those method names are _very_ similar, I had to stare for a while to figure out these two tests were different (but, my glasses are currently at the optometrist undergoing a change in lens, so that might be it as well).

File: test/common/event/guarddog_impl_test.cc [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+TEST_F(GuardDogDeathTest, NearDeathTest) {
+  startNeglectedDogAndTouchedDog();
+  SUCCEED(); // if we haven't crashed by now
```
Comment (htuch) [inline test/common/event/guarddog_impl_test.cc]: I think by default you succeed in gtests..

File: test/common/event/guarddog_impl_test.cc [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+TEST_F(GuardDogMissTest, MissTest) {
+  startNeglectedDogAndWait(false);
```
Comment (htuch) [inline test/common/event/guarddog_impl_test.cc]: Prefer code to be in the `TEST_F` unless it needs to be reused between tests. Even then, sometimes you can structure parameterized tests via `TEST_P` when it makes sense (and without jumping through too many hoops). You can access any member variable or method of the test fixture from the `TEST_F`.

File: test/common/event/guarddog_impl_test.cc [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+  testing::NiceMock<MockSystemTimeSource> time_source;
+  GuardDogImpl gd(stats, config, time_source);
+}
+
+TEST(GuardDogBasicTest, LoopIntervalNoKillTest) {
+  testing::NiceMock<Stats::MockStore> stats;
+  testing::NiceMock<Server::Configuration::MockMain> config(40, 50, 0, 0);
+  testing::NiceMock<MockSystemTimeSource> time_source;
+  GuardDogImpl gd(stats, config, time_source);
+  ASSERT_EQ(gd.loopInterval(), 40);
```
Comment (htuch) [inline test/common/event/guarddog_impl_test.cc]: Generally best to use `EXPECT_EQ` unless your checking a condition that will make the test explode, e.g. a nullptr where later it might be used by some other test code.

File: source/common/event/guarddog_impl.cc (hunk 6 above)
Comment (dnoe) [inline source/common/event/guarddog_impl.cc]: I felt that 1/2 might be excessive timer wake ups.  But, it's not like the timer routine here is very complicated... and we're not running Envoy on mobile/power sensitive devices.

File: source/common/event/guarddog_impl.cc (hunk 2 above)
Comment (dnoe) [inline source/common/event/guarddog_impl.cc]: I didn't think that was possible since loop_interval_ is const? But, come to think of it, I could just write it as a lambda that executes and returns the value used in the initializer list, and the lambda could contain a somewhat simpler and easier to read version of this min mess.

File: source/common/event/guarddog_impl.cc (hunk 5 above)
Comment (dnoe) [inline source/common/event/guarddog_impl.cc]: Good idea

File: source/common/event/guarddog_impl.cc (hunk 7 above)
Comment (dnoe) [inline source/common/event/guarddog_impl.cc]: loop_interval_ is const, so I'd say this is OK?  Is there a build target to run tsan like there is for asan?

File: source/common/event/guarddog_impl.h (hunk 10 above)
Comment (dnoe) [inline source/common/event/guarddog_impl.h]: This was mostly a note for myself, I'll clean it up.  The old default is a reference to the hardcoded values from the original non-threaded watchdog stats.

File: source/common/event/guarddog_impl.h (hunk 9 above)
Comment (dnoe) [inline source/common/event/guarddog_impl.h]: Good catch, there's my old style preferences leaking :)

File: test/integration/fake_upstream.cc (hunk 11 above)
Comment (dnoe) [inline test/integration/fake_upstream.cc]: The stats_store_ was only passed through this call chain because it was used by the old watchdog timer.  Moving the stats incrementing activity into GuardDog made these members unused/args not needed, so I've removed them.

File: test/common/event/guarddog_impl_test.cc (hunk 13 above)
Comment (dnoe) [inline test/common/event/guarddog_impl_test.cc]: So, there's a reason why I had to do this for the death test: The thread creation must happen within the statement inside EXPECT_DEATH for it to work properly.  Since that happens at the very beginning of the test it's best as a separate method.  But it would be good to make the name a bit more distinctive/match the testcase name.

File: test/common/event/guarddog_impl_test.cc (hunk 14 above)
Comment (dnoe) [inline test/common/event/guarddog_impl_test.cc]: Yeah, that's true (the doc for SUCCEED says it's just for self-documenting purposes).

File: test/common/event/guarddog_impl_test.cc (hunk 15 above)
Comment (dnoe) [inline test/common/event/guarddog_impl_test.cc]: I think I can keep the code in TEST_F for everything but the actual death tests, I'll do that.

File: test/common/event/guarddog_impl_test.cc (hunk 16 above)
Comment (dnoe) [inline test/common/event/guarddog_impl_test.cc]: OK, thanks

File: include/envoy/event/guarddog.h [hunk 17]
```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+#include "envoy/event/watchdog.h"
+#include "envoy/stats/stats.h"
+
+namespace Event {
```
Comment (dnoe) [inline include/envoy/event/guarddog.h]: I am leaning in that direction as well... @mattklein123 ?

File: source/common/event/guarddog_impl.cc (hunk 7 above)
Comment (htuch) [inline source/common/event/guarddog_impl.cc]: It's probably OK. I have a preference as a reader for wider locked blocks and as performance optimizer for shorter ones. Maybe just add a comment that all shared state accessed is const.

File: test/common/event/guarddog_impl_test.cc (hunk 13 above)
Comment (htuch) [inline test/common/event/guarddog_impl_test.cc]: Can you put this inside a lambda local to the TEST_F?

Comment (dnoe) [inline test/common/event/guarddog_impl_test.cc]: Good idea, I think that will work.

File: include/envoy/server/configuration.h (hunk 1 above)
Comment (dnoe) [inline include/envoy/server/configuration.h]: I don't think doxygen actually does anything special with that, but this is a good convention to follow.  I will update the existing line above (statsFlushInterval()) in the same way.

File: docs/configuration/overview/overview.rst [hunk 18]
```diff
@@ -57,6 +61,24 @@ stats_flush_interval_ms
   performance reasons Envoy latches counters and only flushes counters and gauges at a periodic
   interval. If not specified the default is 5000ms (5 seconds).
 
+watchdog_miss_timeout_ms
+  *(optional, integer)* The time in milliseconds after which we count a nonresponsive thread in the
```
Comment (mattklein123) [inline docs/configuration/overview/overview.rst]: nit: "after which Envoy counts"

File: docs/configuration/overview/overview.rst [hunk 19]
```diff
@@ -57,6 +61,24 @@ stats_flush_interval_ms
...
+  "server.watchdog_miss" statistic. If not specified the default is 200ms.
+
+watchdog_megamiss_timeout_ms
+  *(optional, integer)* The time in milliseconds after which we count a nonresponsive thread in the
```
Comment (mattklein123) [inline docs/configuration/overview/overview.rst]: same

File: docs/configuration/overview/overview.rst [hunk 20]
```diff
@@ -57,6 +61,24 @@ stats_flush_interval_ms
...
+  "server.watchdog_mega_miss" statistic. If not specified the default is 1000ms.
+
+watchdog_kill_timeout_ms
```
Comment (mattklein123) [inline docs/configuration/overview/overview.rst]: My feeling is to default kill behavior to off and let people turn it on as desired. There are testing/dev/virtual envs where I think this could end up firing pretty easily. Thoughts?

File: include/envoy/server/configuration.h [hunk 21]
```diff
@@ -128,10 +128,31 @@ class Main {
...
+   *         The server latches counters.
    */
   virtual std::chrono::milliseconds statsFlushInterval() PURE;
+
+  /**
+   * @return std::chrono::milliseconds the time interval after which we count a nonresponsive thread
+   *         event as a "miss" statistic.
+   */
+  virtual std::chrono::milliseconds wdMissTimeout() const PURE;
+  /**
```
Comment (mattklein123) [inline include/envoy/server/configuration.h]: nit: newlines between functions defs and next comment to match style elsewhere (same below)

File: source/server/guarddog_impl.cc [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+                         multikillEnabled() ? multi_kill_timeout_ : min_of_nonfatal,
+                         min_of_nonfatal});
+      }()),
+      watchdog_miss_counter_(stats_store.counter("server.watchdog_miss")),
+      watchdog_megamiss_counter_(stats_store.counter("server.watchdog_mega_miss")),
+      run_thread_(true) {
+  start();
+}
+
+GuardDogImpl::~GuardDogImpl() { exit(); }
```
Comment (mattklein123) [inline source/server/guarddog_impl.cc]: ? Confused by the exit() call here.

File: source/server/guarddog_impl.h [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+class GuardDogImpl : public GuardDog {
+public:
+  /**
+   * @param stats_store Statistics store to write watchdog_miss and
+   * watchdog_mega_miss events into.
+   * @param config Configuration object.
+   *
+   * See the configuration documentation for details on the timeout settings.
+   */
+  GuardDogImpl(Stats::Store& stats_store, const Server::Configuration::Main& config,
```
Comment (mattklein123) [inline source/server/guarddog_impl.h]: As long as you are refactoring this code, please switch this to State::Scope& scope (new since this code was first written)

File: test/common/event/guarddog_impl_test.cc [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+ * green in the coverage report. However, rest assured from the results of the
+ * test: these lines are in fact covered.
+ */
+class GuardDogDeathTest : public testing::Test {
+protected:
+  GuardDogDeathTest()
+      : config_kill_(1000, 1000, 100, 1000), config_multikill_(1000, 1000, 1000, 500),
+        time_point_(std::chrono::system_clock::now()) {}
+
+  testing::NiceMock<Server::Configuration::MockMain> config_kill_;
```
Comment (mattklein123) [inline test/common/event/guarddog_impl_test.cc]: nit: Can we do using testing::NiceMock above and avoid the extra verbiage here and elsewhere in tests.

File: source/server/watchdog_impl.h [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+   * @param thread_id A system thread ID (such as from Thread::currentThreadId())
+   * @param interval WatchDog timer interval (used after startWatchdog())
+   */
+  WatchDogImpl(int32_t thread_id, SystemTimeSource& tsource, std::chrono::milliseconds interval)
+      : thread_id_(thread_id), time_source_(tsource),
+        latest_touch_time_(tsource.currentSystemTime()), timer_interval_(interval) {}
+
+  void startWatchdog(Event::Dispatcher& dispatcher);
+
+  void touch() { latest_touch_time_.store(time_source_.currentSystemTime()); }
```
Comment (mattklein123) [inline source/server/watchdog_impl.h]: nit: add override specifier where applicable, and put overriden functions beneath comment. E.g.,

// Server::Watchdog
void touch() override { ... }

File: source/server/watchdog_impl.h [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+
+  int32_t threadId() const { return thread_id_; }
+  SystemTime lastTouchTime() const { return latest_touch_time_.load(); }
+
+private:
+  int32_t thread_id_;
```
Comment (mattklein123) [inline source/server/watchdog_impl.h]: const

File: source/server/watchdog_impl.h [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+  SystemTimeSource& time_source_;
+  std::atomic<SystemTime> latest_touch_time_;
+  Event::TimerPtr timer_;
+  std::chrono::milliseconds timer_interval_;
```
Comment (mattklein123) [inline source/server/watchdog_impl.h]: const

File: source/server/guarddog_impl.h [hunk 28]
```diff
@@ -0,0 +1,78 @@
...
+               SystemTimeSource& tsource);
+  ~GuardDogImpl();
+
+  WatchDogSharedPtr createWatchDog(int32_t thread_id) override;
```
Comment (mattklein123) [inline source/server/guarddog_impl.h]: nit: same as below, please put overriden functions last in the visiblity section, under comment with name of interface. Makes it easier to see which is override vs. impl function.

File: include/envoy/server/watchdog.h [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+   */
+  virtual void startWatchdog(Event::Dispatcher& dispatcher) PURE;
+
+  /**
+   * Manually indicate that you are still alive by calling this.
+   *
+   * This can be used if this is later used on a thread where there is no dispatcher.
+   */
+  virtual void touch() PURE;
+  virtual int32_t threadId() const PURE;
```
Comment (mattklein123) [inline include/envoy/server/watchdog.h]: What is threadId() actually used for? I might be missing it but I don't see it being used by anything. If not used can we delete, though should we use it to print the stuck thread ID in the panic?

File: source/server/guarddog_impl.cc [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+
+namespace Server {
+
+GuardDogImpl::GuardDogImpl(Stats::Store& stats_store, const Server::Configuration::Main& config,
+                           SystemTimeSource& tsource)
+    : time_source_(tsource), miss_timeout_(config.wdMissTimeout()),
+      megamiss_timeout_(config.wdMegaMissTimeout()), kill_timeout_(config.wdKillTimeout()),
+      multi_kill_timeout_(config.wdMultiKillTimeout()),
+      loop_interval_([&]() -> std::chrono::milliseconds {
+        const auto min_of_nonfatal = std::min(miss_timeout_, megamiss_timeout_);
```
Comment (mattklein123) [inline source/server/guarddog_impl.cc]: This pretty dense. Can we get a comment on what this is doing.

File: source/server/guarddog_impl.cc [hunk 31]
```diff
@@ -0,0 +1,103 @@
...
+    watched_dogs_.push_back(new_watchdog);
+  }
+  new_watchdog->touch();
+  return new_watchdog;
+}
+
+void GuardDogImpl::stopWatching(WatchDogSharedPtr wd) {
+  std::lock_guard<std::mutex> guard(wd_lock_);
+  auto found_wd = std::find(watched_dogs_.begin(), watched_dogs_.end(), wd);
+  if (found_wd != watched_dogs_.end()) {
```
Comment (mattklein123) [inline source/server/guarddog_impl.cc]: should this ever actually happen? Can we at least ASSERT() it was found?

File: source/server/guarddog_impl.cc (hunk 22 above)
Comment (mattklein123) [inline source/server/guarddog_impl.cc]: Oops, sorry, from quick read I thought this was exit(0) :)

Can we rename this function. :)

File: test/common/event/guarddog_impl_test.cc (hunk 24 above)
Comment (dnoe) [inline test/common/event/guarddog_impl_test.cc]: Sorry, I didn't follow this: what do you want to use NiceMock for?

Comment (dnoe) [inline test/common/event/guarddog_impl_test.cc]: Agh, nevermind... "using" C++ keyword, not general english word.  Yes, good idea.

File: docs/configuration/overview/overview.rst (hunk 20 above)
Comment (dnoe) [inline docs/configuration/overview/overview.rst]: I think leaving the kill behavior off by default is OK for now.  It's easy to change the defaults later or just use a custom config.

File: source/server/guarddog_impl.cc (hunk 22 above)
Comment (dnoe) [inline source/server/guarddog_impl.cc]: Hah.  stop() is a better name, especially with start() already.

File: source/server/guarddog_impl.cc (hunk 31 above)
Comment (dnoe) [inline source/server/guarddog_impl.cc]: No, it's a harmless case though.  If for some reason this method was called twice before the shared pointer goes out of scope the second call will not find the WatchDog and then when the shared pointer goes out of scope all resources are cleaned up, there's no chance of a crash or resource leak.  However, calling it twice sounds indicative of some other problem so ASSERT seems to make sense.  It looks like in prod builds ASSERTs are compiled out so I'll structure the change so that we do nothing harmlessly when ASSERTs are disabled.

File: source/server/guarddog_impl.h (hunk 23 above)
Comment (dnoe) [inline source/server/guarddog_impl.h]: I think you mean Stats::Scope& which looks like an updated version of Stats::Store& ?  Is there anything special that needs to be done with regards to the thread local storage mentioned here since this is incrementing the counter in a new thread?

File: include/envoy/server/watchdog.h (hunk 29 above)
Comment (dnoe) [inline include/envoy/server/watchdog.h]: I was planning to put it into the PANIC message but I hadn't discovered the fmt::format until just now.  I'll use that to put the thread ID in panic message.

File: source/server/guarddog_impl.h (hunk 23 above)
Comment (mattklein123) [inline source/server/guarddog_impl.h]: Sorry yes typo: Stats::Scope. No, the counters that are allocated are thread safe. Eventually we will be passing Stats::Scope everywhere and keep the Store to the server. There should be no other changes needed to your code.

File: source/server/guarddog_impl.cc [hunk 32]
```diff
@@ -0,0 +1,112 @@
...
+    watched_dogs_.erase(found_wd);
+  } else {
+    ASSERT(false);
+  }
+}
+
+bool GuardDogImpl::waitOrDetectStop() {
+  force_checked_event_.notify_all();
+  std::lock_guard<std::mutex> guard(exit_lock_);
```
Comment (htuch) [inline source/server/guarddog_impl.cc]: Observation: we use `std::lock_guard` here and `std::unique_lock` in the integration tests. Should we just standardize on the `std::lock_guar`d unless we need move semantics? No action required.

File: test/server/guarddog_impl_test.cc [hunk 33]
```diff
@@ -0,0 +1,156 @@
...
+    GuardDogImpl gd(fakestats_, config_multikill_, time_source_);
+    auto unpet_dog1 = gd.createWatchDog(0);
+    gd.forceCheckForTest();
+    auto unpet_dog2 = gd.createWatchDog(1);
+    gd.forceCheckForTest();
+    time_point_ += std::chrono::milliseconds(501);
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    gd.forceCheckForTest();
+  };
+  EXPECT_DEATH(die_function(), "");
```
Comment (htuch) [inline test/server/guarddog_impl_test.cc]: Shouldn't only the last `gd.forceCheckForTest()` be under the `EXPECT_DEATH`? In addition to stating that we die, we want to know we don't die if only one thread exceeds the threshold. Same applies, lesser extent to the previous test as well.

File: test/server/guarddog_impl_test.cc [hunk 34]
```diff
@@ -0,0 +1,156 @@
...
+  auto pet_dog = gd.createWatchDog(1);
+  for (int i = 0; i < 6; i++) {
+    time_point_ += std::chrono::milliseconds(100);
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    pet_dog->touch();
+    gd.forceCheckForTest();
+  }
+}
+
+TEST_F(GuardDogMissTest, MissTest) {
```
Comment (htuch) [inline test/server/guarddog_impl_test.cc]: It might be helpful to the reader to have a short english explanation of what the test does above. E.g. "This test validates that when we miss that stats are incremented."

File: test/server/guarddog_impl_test.cc [hunk 35]
```diff
@@ -0,0 +1,156 @@
...
+  EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+  GuardDogImpl gd(stats_store, config_miss, time_source_);
+  auto unpet_dog = gd.createWatchDog(0);
+  time_point_ += std::chrono::milliseconds(501);
+  EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+  gd.forceCheckForTest();
+  gd.stopWatching(unpet_dog);
+  unpet_dog = nullptr;
```
Comment (htuch) [inline test/server/guarddog_impl_test.cc]: Should there be a check for stats updated?

File: test/server/guarddog_impl_test.cc [hunk 36]
```diff
@@ -0,0 +1,156 @@
...
+
+TEST_F(GuardDogMissTest, MegaMissTest) {
+  EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+  GuardDogImpl gd(stats_store, config_mega, time_source_);
+  auto unpet_dog = gd.createWatchDog(0);
+  time_point_ += std::chrono::milliseconds(501);
+  EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+  gd.forceCheckForTest();
+  gd.stopWatching(unpet_dog);
+  unpet_dog = nullptr;
```
Comment (htuch) [inline test/server/guarddog_impl_test.cc]: Ditto.

File: test/server/guarddog_impl_test.cc [hunk 37]
```diff
@@ -0,0 +1,156 @@
...
+  gd.stopWatching(watched_dog);
+}
+// If this test fails it is because the SystemTime object has become nontrivial
+// or we are compiling under a compiler and library combo that makes the
+// SystemTime object require a lock to be atomicly modified.
+//
+// The WatchDog/GuardDog relies on this being a lock free atomic for perf
+// reasons so some workaround will be required if this test starts failing.
+TEST(WatchDogTimeTest, AtomicIsAtomicTest) {
+  ASSERT(std::atomic<ProdSystemTimeSource>{}.is_lock_free());
```
Comment (htuch) [inline test/server/guarddog_impl_test.cc]: Is it the `ProdSystemTimeSource` or `SystemTime` you want to be atomic lock free here?

File: test/server/guarddog_impl_test.cc (hunk 33 above)
Comment (dnoe) [inline test/server/guarddog_impl_test.cc]: No, this is the quirk of the death test when there are threads in play: The thread must be started within the statement in EXPECT_DEATH.

File: test/server/guarddog_impl_test.cc (hunk 37 above)
Comment (dnoe) [inline test/server/guarddog_impl_test.cc]: Good catch! yes, it's the SystemTime.

File: test/server/guarddog_impl_test.cc [hunk 38]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+TEST_F(GuardDogDeathTest, NearDeathTest) {
+  InSequence s;
+  EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+  GuardDogImpl gd(fakestats_, config_multikill_, time_source_);
+  auto unpet_dog = gd.createWatchDog(0);
+  auto pet_dog = gd.createWatchDog(1);
+  for (int i = 0; i < 6; i++) {
+    time_point_ += std::chrono::milliseconds(100);
```
Comment (htuch) [inline test/server/guarddog_impl_test.cc]: It's hard to tell from reading this how close to death we are in this test - is it 600ms vs. 1000ms? Could that be somehow made clearer?

File: test/server/guarddog_impl_test.cc (hunk 33 above)
Comment (htuch) [inline test/server/guarddog_impl_test.cc]: Should you then have a separate test that is identical, except for the final `forceCheckForTest` to get this additional behavioral coverage?

File: test/server/guarddog_impl_test.cc (hunk 35 above)
Comment (dnoe) [inline test/server/guarddog_impl_test.cc]: Ideally. I couldn't figure out how to do this with the stats mocking.  Have any pointers for somewhere else where it's done?

File: source/server/guarddog_impl.cc (hunk 32 above)
Comment (dnoe) [inline source/server/guarddog_impl.cc]: I wasn't previously familiar with unique_lock.  I think I'd be in favor of using the thinner lock_guard unless the additional features of unique_lock are required.

File: test/server/guarddog_impl_test.cc (hunk 35 above)
Comment (mattklein123) [inline test/server/guarddog_impl_test.cc]: There are multiple ways of doing stats testing. You can actually use a MockStore, but in general that's kind of a pain so we don't do that. If you look through the tests you will see that we just validate that a particular counter/gauge within the store is at the right value. For example ```EXPECT_EQ(1UL, store.counter("foo").value())```

There are tons of examples of this throughout the code. And yes, please validate stats if possible.

File: test/server/guarddog_impl_test.cc (hunk 33 above)
Comment (dnoe) [inline test/server/guarddog_impl_test.cc]: Good idea, it will create some more "near death" cases.

File: test/server/guarddog_impl_test.cc (hunk 35 above)
Comment (dnoe) [inline test/server/guarddog_impl_test.cc]: OK, I'll just use real stats and check for the result rather than mock and check for activity.  I was trying to figure out how to do it with MockStore and it was indeed proving painful.

Comment (mattklein123) [inline test/server/guarddog_impl_test.cc]: See this PR tests for examples: https://github.com/lyft/envoy/pull/656/files

File: source/server/guarddog_impl.h [hunk 39]
```diff
@@ -0,0 +1,79 @@
...
+ * It launches a thread that scans at an interval the minimum of the configured
+ * intervals.  If it finds starved threads or suspected deadlocks it will take
+ * the appropriate action depending on the config parameters described below.
+ *
+ * Thread lifetime is tied to GuardDog object lifetime (RAII style).
+ */
+class GuardDogImpl : public GuardDog {
+public:
+  /**
+   * @param stats_store Statistics store to write watchdog_miss and
```
Comment (mattklein123) [inline source/server/guarddog_impl.h]: nit: stats_scope

File: test/server/guarddog_impl_test.cc [hunk 40]
```diff
@@ -0,0 +1,225 @@
+#include "common/common/utility.h"
+#include "common/stats/stats_impl.h"
+
+#include "envoy/common/time.h"
```
Comment (mattklein123) [inline test/server/guarddog_impl_test.cc]: nit: envoy headers come first (move to top)
